File: scripts/recipes/production_order_runbook.py
"""
Production Order Runbook -- Create to Invoice (checklist + one automated stage)

The runbook page is a checklist, not a script. This file prints that
checklist (stages 1-2 and 4-7: what to call, the fields that matter, and the
trap for each), and automates the stage most people automate first --
Stage 3: generate the production pick ticket at the STOCK location with
m_picktickets (creates the ticket record AND returns the PDF), then read the
new ticket back with POST /api/v2/transaction/get to check its status.

Mirrors: docs/recipes/production-order-runbook.md

Usage:
    python scripts/recipes/production_order_runbook.py            # checklist + dry run
    python scripts/recipes/production_order_runbook.py --execute  # run Stage 3 + read-back
"""

import argparse
import base64
import json
import re
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent))

import httpx
from common.auth import get_token, get_auth_headers, get_ui_server_url
from common.config import load_config

import warnings
warnings.filterwarnings("ignore")

# --- Configuration (generic placeholders -- substitute your own) ------------
PROD_ORDER = "1000123"   # production order number
STOCK_LOCATION = "10"    # where the components stock (NOT necessarily the make location)

CHECKLIST = """\
Runbook stages (this script automates Stage 3 only -- see the recipe page
docs/recipes/production-order-runbook.md and the linked recipes for the rest):

Stage 1 - Create the production order
  Path A: sales order auto-create (make-to-order) -- enter the assembly line
          via the Interactive API (scripts/recipes/order_with_assembly.py).
          Auto-create NETS AGAINST STOCK: on-hand means no order spawns.
  Path B: direct build-to-stock -- ProductionOrder window: header
          source_loc_id, then TABPAGE_17.tp_17_dw_17 assembly_item_id +
          qty_to_make.
  Traps:  salesrep must be valid at the sales location; order date must
          differ from required date.

Stage 2 - Log labor BEFORE printing (scripts/recipes/record_labor_time.py)
  Trap:   labor added after printing (no reprint) sits on no ticket
          (qty_on_pick_tickets = 0) and completion fails with
          "components have a quantity used of 0."

Stage 3 - Print the pick ticket and form  [AUTOMATED BELOW]
  m_picktickets at POST /api/v2/process/pdfreport, at the STOCK location.
  Traps:  print_pick_ticket on a ProductionOrder transaction emits only at
          the MAKE location; never post an m_* report to /api/v2/transaction
          (it returns Succeeded and emits nothing); the order's form must
          already be printed (prod_order_hdr.printed = 'Y').

Stage 4 - Confirm the pick (Interactive API ONLY)
  ProductionOrderPicking window: load the ticket on
  TP_PRODPICKTICKETCONF.tp_prodpickticketconf (key prod_pick_ticket_number),
  set row_status_flag = "Confirm", save. Confirm EVERY ticket -- parts AND
  labor/intangibles.
  Trap:   a bare Transaction API confirm is a SHELL confirm -- status flips
          to 1962 but qty_applied stays 0 and no stock moves.

Stage 5 - Complete the order (ProductionOrderProcessing window)
  Select the line on TABPAGE_17.tp_17_dw_17, set qty_to_complete; then on
  TABPAGE_ASSEMBLY_BIN.tabpage_assembly_bin set bin_cd and unit_quantity as
  TWO SEPARATE change calls; optional new_cost on TABPAGE_18.tp_18_dw_18; save.
  Trap:   combining bin_cd and unit_quantity in one call drops the quantity.

Stage 6 - Ship and invoice the linked sales order
  Order transaction with print_tix = ON on TP_FRONTCOUNTER.tp_frontcounter,
  then the Shipping service keyed by pick_ticket_no -- retrieve and save
  (create_invoice defaults ON: the save ships AND invoices).
  Traps:  the item needs a packaging code; for contract pricing leave
          unit_price unset.

Stage 7 - Fix quantity fallout (scripts/recipes/inventory_adjustment.py)

Ticket status codes (prod_pick_ticket_hdr.row_status_flag):
  702 Open / 1962 Confirmed / 1268 Completed
  A 1962 alone does not prove stock moved (see shell confirm).
"""


def build_report_payload() -> dict:
    """Build the Stage 3 m_picktickets payload for PROD_ORDER at STOCK_LOCATION.

    Returns:
        dict: Complete pdfreport payload (numeric Status/Type 0, Keys []).
    """
    return {
        "Name": "m_picktickets",
        "UseCodeValues": True,  # m_picktickets REQUIRES code values; False returns HTTP 500
        "Transactions": [{
            "Status": 0,        # reports use numeric 0, not "New"
            "DataElements": [{
                "Keys": [],
                "Type": 0,
                "Name": "TABPAGE_1.tp_1_dw_1",
                "Rows": [{"Edits": [
                    {"Name": "create_pick_ticket_type", "Value": "P"},  # code "P" = Production Order
                    {"Name": "beg_prod_order", "Value": PROD_ORDER},
                    {"Name": "end_prod_order", "Value": PROD_ORDER},
                    {"Name": "location_id", "Value": STOCK_LOCATION},
                ]}],
            }],
        }],
    }


def generate_pick_ticket(ui_server: str, headers: dict, verify_ssl: bool) -> str:
    """Run m_picktickets, save the PDF, and return the saved file name.

    Args:
        ui_server: UI server URL.
        headers: Auth headers.
        verify_ssl: Whether to verify SSL certificates.

    Returns:
        str: The saved PDF file name (contains the ticket number).

    Raises:
        SystemExit: When the report fails.
    """
    resp = httpx.post(
        f"{ui_server}/api/v2/process/pdfreport",  # NOT /api/v2/transaction (silent no-op there)
        headers=headers, json=build_report_payload(), verify=verify_ssl,
        follow_redirects=True, timeout=120,
    )
    resp.raise_for_status()
    result = resp.json()

    if not isinstance(result, list):  # errors come back as an envelope, not an array
        raise SystemExit(f"Report failed: {result.get('ErrorMessage')}")

    doc = result[0]
    if doc["ResponseStatus"]["StatusCode"] != "Success" or not doc.get("DocumentData"):
        raise SystemExit(f"Report failed: {doc['ResponseStatus'].get('Message')}")

    file_name = doc["FileName"]  # e.g. "PPT123456 PRODUCTION_PICK_TICKET.pdf"
    with open(file_name, "wb") as f:
        f.write(base64.b64decode(doc["DocumentData"]))
    print(f"Saved {file_name}")
    return file_name


def read_ticket_status(ui_server: str, headers: dict, verify_ssl: bool,
                       ticket_no: str) -> None:
    """Read the new ticket back via /transaction/get and print its status.

    Args:
        ui_server: UI server URL.
        headers: Auth headers.
        verify_ssl: Whether to verify SSL certificates.
        ticket_no: Production pick-ticket number (from the PDF FileName).
    """
    get_payload = {
        "ServiceName": "ProductionOrderPicking",
        "TransactionStates": [{
            "DataElementName": "TP_PRODPICKTICKETCONF.tp_prodpickticketconf",
            "Keys": [{"Name": "prod_pick_ticket_number", "Value": ticket_no}],
        }],
    }
    resp = httpx.post(f"{ui_server}/api/v2/transaction/get",
                      headers=headers, json=get_payload, verify=verify_ssl,
                      follow_redirects=True, timeout=60)
    resp.raise_for_status()

    for txn in resp.json().get("Transactions", []):
        for de in txn.get("DataElements", []):
            for row in de.get("Rows", []):
                fields = {e["Name"]: e["Value"] for e in row.get("Edits", [])}
                if "row_status_flag" in fields:
                    # 702 = Open, 1962 = Confirmed, 1268 = Completed
                    print(f"Ticket {fields.get('prod_pick_ticket_number')} "
                          f"for prod order {fields.get('prod_order_number')}: "
                          f"status {fields.get('row_status_flag')}")


def main() -> None:
    """Entry point: print the runbook checklist; on --execute run Stage 3."""
    parser = argparse.ArgumentParser(
        description="Production order runbook (docs/recipes/production-order-runbook.md)")
    parser.add_argument("--execute", action="store_true",
                        help="Run Stage 3 (m_picktickets) -- CREATES the pick-ticket "
                             "record in P21 (default: dry run)")
    args = parser.parse_args()

    print("Recipe - Production Order Runbook (Create to Invoice)")
    print("=" * 60)
    print()
    print(CHECKLIST)

    if not args.execute:
        print("DRY RUN - Stage 3 payload that would be POSTed to "
              "{ui_server}/api/v2/process/pdfreport:")
        print(json.dumps(build_report_payload(), indent=2))
        print("\nRe-run with --execute to generate the pick ticket and read its status back.")
        return

    config = load_config()
    token_data = get_token(config)
    headers = get_auth_headers(token_data["AccessToken"])
    ui_server = get_ui_server_url(config.base_url, token_data["AccessToken"], config.verify_ssl)
    print(f"UI Server: {ui_server}")

    # --- Stage 3, part 1: generate the pick ticket (record + PDF) ---
    file_name = generate_pick_ticket(ui_server, headers, config.verify_ssl)

    # --- Stage 3, part 2: read the new ticket back (number from the FileName) ---
    match = re.match(r"PPT(\d+)", file_name)
    if not match:
        raise SystemExit(f"Could not extract ticket number from '{file_name}'")
    read_ticket_status(ui_server, headers, config.verify_ssl, match.group(1))


if __name__ == "__main__":
    main()

File: examples/csharp/Recipes/ProductionOrderRunbook.cs
// Recipe: Production Order Runbook — Create to Invoice
// Docs:   docs/recipes/production-order-runbook.md
//
// The recipe page is a CHECKLIST, not a script: create the production order,
// log labor, print the pick ticket(s), confirm the pick, complete (receive)
// the order, and ship + invoice the linked sales order. This example prints
// that checklist as guidance and automates the one stage most people script
// first: generating the pick ticket with m_picktickets and reading the new
// ticket's status back via POST /api/v2/transaction/get.
//
// Key traps (all verified live — see the recipe page):
//   - Shell confirm: confirming a pick via a bare Transaction POST flips the
//     status to 1962 but qty_applied stays 0 and NO stock moves — confirm
//     through the Interactive API ProductionOrderPicking window.
//   - Labor must land on a pick ticket before completion.
//   - bin_cd and unit_quantity are two SEPARATE change calls at completion.
//   - Reports go to /api/v2/process/pdfreport, never /api/v2/transaction.
//
// Uses a raw HttpClient (Common P21Config/P21Auth) because pdfreport is not
// wrapped by P21Client.

using System.Text;
using System.Text.RegularExpressions;
using Newtonsoft.Json.Linq;
using static P21Examples.Recipes.RecipeHelpers;

namespace P21Examples.Recipes;

public static class ProductionOrderRunbook
{
    private const string ProdOrder = "1000123";     // production order number
    private const string StockLocation = "10";      // where the components stock (NOT necessarily the make location)

    public static async Task RunAsync()
    {
        Console.WriteLine("Recipe - Production Order Runbook (Create to Invoice)");
        Console.WriteLine(new string('=', 60));

        PrintChecklist();

        // ------------------------------------------------------------------
        // Automated stage: print the pick ticket and read back its status
        // ------------------------------------------------------------------
        Console.WriteLine("\nAutomated stage — Stage 3: generate the production pick ticket");
        Console.WriteLine(new string('-', 60));

        var report = new JObject
        {
            ["Name"] = "m_picktickets",
            ["UseCodeValues"] = true,  // m_picktickets REQUIRES code values; false returns HTTP 500
            ["Transactions"] = new JArray
            {
                new JObject
                {
                    ["Status"] = 0,    // reports use numeric 0, not "New"
                    ["DataElements"] = new JArray
                    {
                        new JObject
                        {
                            ["Keys"] = new JArray(),
                            ["Type"] = 0,
                            ["Name"] = "TABPAGE_1.tp_1_dw_1",
                            ["Rows"] = new JArray
                            {
                                new JObject
                                {
                                    ["Edits"] = new JArray
                                    {
                                        Edit("create_pick_ticket_type", "P"), // code "P" = Production Order
                                        Edit("beg_prod_order", ProdOrder),
                                        Edit("end_prod_order", ProdOrder),
                                        Edit("location_id", StockLocation),
                                    }
                                }
                            }
                        }
                    }
                }
            }
        };

        PrintPayload("m_picktickets payload (POST /api/v2/process/pdfreport)", report);
        Console.WriteLine(
            "\nNOTE: this WRITES — it creates the pick-ticket record at the stock" +
            " location in addition to returning the PDF.\n" +
            "Prerequisite: prod_order_hdr.printed = 'Y' (ProductionOrder transaction" +
            " with print_form = ON).");

        if (!ConfirmExecute())
            return;

        var (rawHttp, uiServer, _) = await CreateRawClientAsync();
        using var http = rawHttp;

        // --- 1. Generate the pick ticket (creates the record + returns the PDF) ---
        // NOT /api/v2/transaction (silent no-op there)
        var reportResp = await http.PostAsync(
            $"{uiServer}/api/v2/process/pdfreport",
            new StringContent(report.ToString(), Encoding.UTF8, "application/json"));
        var reportText = await reportResp.Content.ReadAsStringAsync();
        if (!reportResp.IsSuccessStatusCode)
        {
            Console.WriteLine($"HTTP {(int)reportResp.StatusCode}: {reportText}");
            return;
        }

        var reportBody = JToken.Parse(reportText);
        if (reportBody.Type != JTokenType.Array)  // errors come back as an envelope, not an array
        {
            Console.WriteLine($"Report failed: {reportBody["ErrorMessage"]}");
            return;
        }

        var doc = (JObject)((JArray)reportBody)[0];
        if (doc["ResponseStatus"]?["StatusCode"]?.ToString() != "Success" ||
            string.IsNullOrEmpty(doc["DocumentData"]?.ToString()))
        {
            Console.WriteLine($"Report failed: {doc["ResponseStatus"]?["Message"]}");
            return;
        }

        var fileName = doc["FileName"]!.ToString(); // e.g. "PPT123456 PRODUCTION_PICK_TICKET.pdf"
        await File.WriteAllBytesAsync(
            fileName, Convert.FromBase64String(doc["DocumentData"]!.ToString()));
        Console.WriteLine($"Saved {fileName}");

        // --- 2. Verify: read the new ticket back (number comes from the FileName) ---
        var match = Regex.Match(fileName, @"PPT(\d+)");
        if (!match.Success)
        {
            Console.WriteLine($"Could not parse a ticket number from '{fileName}'");
            return;
        }
        var ticketNo = match.Groups[1].Value;

        Console.WriteLine($"\nVerify: reading ticket {ticketNo} back via /transaction/get:");
        Console.WriteLine(new string('-', 50));

        var getPayload = new JObject
        {
            ["ServiceName"] = "ProductionOrderPicking",
            ["TransactionStates"] = new JArray
            {
                new JObject
                {
                    ["DataElementName"] = "TP_PRODPICKTICKETCONF.tp_prodpickticketconf",
                    ["Keys"] = new JArray
                    {
                        new JObject { ["Name"] = "prod_pick_ticket_number", ["Value"] = ticketNo }
                    }
                }
            }
        };

        var getResp = await http.PostAsync(
            $"{uiServer}/api/v2/transaction/get",
            new StringContent(getPayload.ToString(), Encoding.UTF8, "application/json"));
        getResp.EnsureSuccessStatusCode();
        var getResult = JObject.Parse(await getResp.Content.ReadAsStringAsync());

        foreach (var txn in getResult["Transactions"] as JArray ?? new JArray())
        foreach (var de in txn["DataElements"] as JArray ?? new JArray())
        foreach (var row in de["Rows"] as JArray ?? new JArray())
        {
            var fields = new Dictionary<string, string>();
            foreach (var edit in row["Edits"] as JArray ?? new JArray())
                fields[edit["Name"]!.ToString()] = edit["Value"]?.ToString() ?? "";
            if (fields.ContainsKey("row_status_flag"))
                // 702 = Open, 1962 = Confirmed, 1268 = Completed
                Console.WriteLine(
                    $"  Ticket {fields.GetValueOrDefault("prod_pick_ticket_number")} " +
                    $"for prod order {fields.GetValueOrDefault("prod_order_number")}: " +
                    $"status {fields.GetValueOrDefault("row_status_flag")} " +
                    "(702 Open / 1962 Confirmed / 1268 Completed)");
        }
    }

    private static void PrintChecklist()
    {
        Console.WriteLine(@"
Runbook checklist (each stage links to a deep dive in the recipe page):

  Stage 1 - Create the production order
    Path A: sales order auto-create (Interactive API when the line must
            explode — see OrderWithAssembly). Auto-create NETS against
            stock: on-hand means no production order spawns.
    Path B: direct build-to-stock via the ProductionOrder window
            (header source_loc_id; TABPAGE_17.tp_17_dw_17
            assembly_item_id + qty_to_make).
    Traps:  salesrep must be valid at the sales location; order date
            must differ from required date.

  Stage 2 - Log labor BEFORE printing (see RecordLaborTime)
    Trap:   labor added after printing (no reprint) sits at
            qty_on_pick_tickets = 0 and completion fails with
            ""components have a quantity used of 0"".

  Stage 3 - Print the pick ticket and form  [AUTOMATED BELOW]
    ProductionOrder transaction (print_pick_ticket/print_form = ON) or
    m_picktickets at POST /api/v2/process/pdfreport.
    Traps:  print_pick_ticket emits only at the MAKE location; never
            post m_* reports to /api/v2/transaction (silent no-op).

  Stage 4 - Confirm the pick (Interactive API ONLY)
    ProductionOrderPicking window, header key prod_pick_ticket_number,
    set row_status_flag = ""Confirm"", save. Confirm EVERY ticket.
    Trap:   a bare Transaction POST produces a SHELL confirm — status
            1962 but qty_applied = 0 and no stock moves.

  Stage 5 - Complete the order (production receipt)
    ProductionOrderProcessing window: qty_to_complete on the line, then
    bin_cd and unit_quantity on TABPAGE_ASSEMBLY_BIN as TWO SEPARATE
    change calls; optional new_cost per component; save (inv_tran PROP).
    Trap:   combining bin_cd and unit_quantity in one call drops the
            quantity.

  Stage 6 - Ship + invoice the linked sales order
    Order transaction with print_tix = ON (creates oe_pick_ticket), then
    the Shipping service keyed by pick_ticket_no — retrieve and save
    (create_invoice defaults ON, so the save ships AND invoices).
    Traps:  the item needs a packaging code; leave unit_price unset for
            contract pricing.

  Stage 7 - Fix quantity fallout (see InventoryAdjustment)

  Cost model: PROP receipt cost = components + labor posted BEFORE
  completion; shipment COGS is the moving average at ship time.");
    }
}
